Quic natsupp (#171)
quicwg/ops-drafts: draft-ietf-quic-manageability.md

endpoints.

{{?RFC4787}} contains some guidance on building NATs to interact constructively
with a wide range of applications. This section extends the discussion to QUIC.

By using the CID, QUIC connections can survive NAT rebindings as long as no
routing function in the path is dependent on client IP address and port to
deliver packets between server and NAT. Reducing the timeout on UDP NATs might
be tempting in light of this property, but not all QUIC server deployments will
be robust to rebinding.

### Resource Conservation

NATs sometimes hit an operational limit where they exhaust available public IP
addresses and ports, and must evict flows from their address/port mapping. CIDs
might appear to offer a way to multiplex many connections over a single
address and port.

However, QUIC endpoints may negotiate new connection IDs inside
cryptographically protected packets, and begin using them at will. Imagine two
clients behind a NAT that are sharing the same public IP address and port. The
NAT is differentiating them using the incoming Connection ID. If one client
secretly changes its connection ID, there will be no mapping for the NAT, and
the connection will suddenly break.

> While mid-connection failure in some cases may seem superior to rejecting QUIC
> outright, HTTP/3 over QUIC falls back to TCP, and
> {{?I-D.ietf-quic-applicability}} recommends TCP fallback for other protocols
> that use QUIC, this is preferable to a connection suddenly black holing and
> timing out, as the recovery time is much longer for a black hole event.

Reviewer: ```suggestion
Some applications might prefer to risk mid-connection failure rather than
not being able to use QUIC. HTTP/3 prefers reliability, recommending that
TCP be used as a fallback and {{?I-D.ietf-quic-applicability}} recommends
TCP fallback for other protocols on the basis that this is preferable to
sudden connection errors and time outs.
```

I had trouble with the phrasing here and I didn't like the comma-splice much.

Reviewer 2: Didn't this completely change the meaning of the first sentence? I don't think it's true that any application would prefer mid-connection failure. The original sentence said "seem superior", meaning that it is not.

Reviewer: "seems" is not a good way to phrase a negative; I read it in the positive.  I was trying for a more or less direct translation of the intent.  If you ask me, I think that you could be more assertive about this and say that 

> QUIC is deliberately designed to fail rather than persist when the network cannot support its operation.  For HTTP/3, this extends to recommending a fallback to TCP-based versions of HTTP rather than persisting with a QUIC connection that might be unstable.

routing function in the path is dependent on client IP address and port to
deliver packets between server and NAT. Reducing the timeout on UDP NATs might
be tempting in light of this property, but not all QUIC server deployments will
be robust to rebinding.

### Resource Conservation

NATs sometimes hit an operational limit where they exhaust available public IP
addresses and ports, and must evict flows from their address/port mapping. CIDs
might appear to offer a way to multiplex many connections over a single
address and port.

However, QUIC endpoints may negotiate new connection IDs inside
cryptographically protected packets, and begin using them at will. Imagine two
clients behind a NAT that are sharing the same public IP address and port. The
NAT is differentiating them using the incoming Connection ID. If one client
secretly changes its connection ID, there will be no mapping for the NAT, and
the connection will suddenly break.

While mid-connection failure in some cases may seem superior to rejecting QUIC
outright, HTTP/3 over QUIC falls back to TCP, and
{{?I-D.ietf-quic-applicability}} recommends TCP fallback for other protocols

> that use QUIC . This is preferable to a connection suddenly black holing and

Reviewer 2: ```suggestion
that use QUIC, this is preferable to a connection suddenly black holing and
```

those addresses and using ports to distinguish those connections. The new
address can also guarantee that packets move through a proxy throughout the life
of a connection, so that the connection can continue with the required state at
that proxy.

This section uses the colloquial term NAT to mean NAPT (section 2.2 of
{{?RFC3022}}), which overloads several IP addresses to one IP address or to an
IP address pool, as commonly deployed in carrier-grade NATs or residential NATs.
QUIC Connection IDs are opaque byte fields that are expressed consistently
across all QUIC versions {{QUIC-INVARIANTS}}. This feature may appear to present
opportunities to optimize NAT port usage and simplify the work of the QUIC
server. In fact, NAT behavior that relies on CID may instead cause connection
failure when endpoints change Connection ID, and disable important protocol
security features.

The remainder of this section explains how QUIC supports NATs better than other
connection-oriented protocols, why NAT use of Connection ID might appear
attractive, and how NAT use of CID can create serious problems for the
endpoints.

> The conclusion of this section is that NATs should retain their existing

Reviewer 3: I'd maybe move this up, or summarize it in the first para. This section should be very clear about the "don't do it"ness of messing with CIDs.

4-tuple-based operation and refrain from parsing or otherwise using QUIC
connection IDs.

{{?RFC4787}} contains some guidance on building NATs to interact constructively
with a wide range of applications. This document extends the discussion to QUIC.

### QUIC and NAT Rebinding

By using the CID, QUIC connections can survive NAT rebindings as long as no
routing function in the path is dependent on client IP address and port to
deliver packets between server and NAT. Reducing the timeout on UDP NATs might
be tempting in light of this property, but not all QUIC server deployments will
be robust to rebinding.

### The Lure of the Connection ID

> There are a few reasons that CID-aware NATs could seemingly appear attractive.

Reviewer 3: editorial, like the comment above: please end this with a snarky conclusion that they are all wrong. :)

Author: Made obsolete by Mirja's suggestion

Author: overcome by events

> There are a few reasons that CID-aware NATs could seemingly appear attractive.

Author: This text is gone

#### Resource Conservation

NATs sometimes hit an operational limit where they exhaust available public IP
addresses and ports, and must evict flows from their address/port mapping. CIDs

> offer a way to multiplex many connections over a single address and port.

Reviewer 3: ```suggestion
might appear to offer a way to multiplex many connections over a single
address and port.
```

However, QUIC endpoints may negotiate new connection IDs inside
cryptographically protected packets, and begin using them at will. Imagine two
clients behind a NAT that are sharing the same public IP address and port. The
NAT is differentiating them using the incoming Connection ID. If one client
secretly changes its connection ID, there will be no mapping for the NAT, and
the connection will suddenly break.

While mid-connection failure in some cases may seem superior to rejecting QUIC

> outright, HTTP/3 over QUIC falls back to TCP. This is preferable to a

Reviewer 3: also note that applicability recommends fallback for non H3 protocols.

connection suddenly black holing and timing out. Furthermore, wide deployment
of NATs with this behavior would make it risky to change Connection IDs in the

> internet, which would thwart various important protocol properties.

Author: Do you have a recommendation?

> internet, which would thwart various important protocol properties.

Reviewer 2: Maybe s/"make it risky to change Connection IDs"/"hinders the use of QUIC's migration function which relies on the ability to change the connection ID any time during the life time of a QUIC connection"...

It is possible, in principle, to encode the client's identity in a connection ID

> using {{QUIC_LB}} and explicit coordination with the NAT. However, QUIC-LB makes assumptions about endpoint mobility and common configuration in server

Reviewer 3: ```suggestion
using {{QUIC_LB}} and explicit coordination with the NAT.
However, QUIC-LB makes assumptions about endpoint mobility and
common configuration in server
```

Reviewer 3: 80-char limit

infrastructure that are almost never valid in client/NAT architectures.
Deploying such a system would include the administrative overhead while not
solving the problem described in this section if the client changes

> networks.

Author: Do you have a suggestion?

> networks.

Reviewer 2: I believe this belonged to the "assumptions about endpoint mobility and common configuration in server
 infrastructure" part. Not fully clear what assumptions you are talking about,

> Note that using connection IDs in this manner would anyway violate the best

Author: Port

common practice to avoid "port overloading" as described in {{?RFC4787}}.

#### "Helping" with routing infrastructure issues

One problem in QUIC deployment is router and switch server infrastructures that
direct traffic based on address-port 4-tuple rather than connection ID. The use
of source IP address means that a NAT rebinding or address migration will
deliver packets to the wrong server. For the reasons described above, routers
and switches will not have access to negotiated CIDs. This is a particular

> problem for low-state load balancers, and a QUIC extension exists {{QUIC_LB}} to

Reviewer 2: Also we have an own section on LB in this draft that we should probably refer to (instead). Maybe we can also just more some of the background infos in this section over there and then have both section right after each other...?!

allow some server-load balancer coordination for routable CIDs.

A NAT at the front of this infrastructure might save the effort of converting
all these devices by decoding routable connection IDs and rewriting the packet
IP addresses to allow consistent routing by legacy devices.

Unfortunately, the change of IP address or port is an important signal to QUIC
endpoints. It requires a review of path-dependent variables like congestion
control parameters. It can also signify various attacks that mislead one

> endpoint about the best peer address for the connection (see section 9 of 

Reviewer 3: ```suggestion
endpoint about the best peer address for the connection (see section 9 of
```

Reviewer 3: \o/ trailing whitespace

{{QUIC-TRANSPORT}}). The QUIC PATH_CHALLENGE and PATH_RESPONSE frames are
intended to detect and mitigate these attacks and verify connectivity to the
new address. This mechanism cannot work if the NAT is bleaching peer address

> changes.

Author: I added another paragraph up front.
